File: android/app/src/main/java/app/cim/family/Crypto.kt
package app.cim.family

import android.util.Base64
import org.bouncycastle.crypto.agreement.X25519Agreement
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.generators.HKDFBytesGenerator
import org.bouncycastle.crypto.params.HKDFParameters
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PublicKeyParameters
import org.json.JSONObject
import java.io.EOFException
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Byte-for-byte the same scheme as modules/family_share/crypto.py on the
 * server: X25519 (ephemeral + static) -> HKDF-SHA256 -> AES-256-GCM, a
 * sealed JSON metadata blob and a framed 1 MiB chunk stream for files.
 */
class CryptoException(msg: String) : Exception(msg)

object Crypto {
    const val VERSION = 1
    const val CHUNK = 1 shl 20
    const val MAX_SKEW = 15 * 60
    private val INFO = "family_share/v1/".toByteArray()
    private val rng = SecureRandom()
    private const val B64 = Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP

    fun b64e(b: ByteArray): String = Base64.encodeToString(b, B64)
    fun b64d(s: String): ByteArray = try { Base64.decode(s, B64) } catch (e: Exception) { throw CryptoException("bad base64") }

    fun generatePrivateKey(): ByteArray = X25519PrivateKeyParameters(rng).encoded
    fun publicKey(priv: ByteArray): ByteArray = X25519PrivateKeyParameters(priv, 0).generatePublicKey().encoded

    fun fingerprint(pub: ByteArray): String {
        val hex = MessageDigest.getInstance("SHA-256").digest(pub).joinToString("") { "%02x".format(it) }.substring(0, 16)
        return hex.chunked(4).joinToString(" ")
    }

    fun sha256Hex(input: InputStream): String {
        val md = MessageDigest.getInstance("SHA-256")
        val buf = ByteArray(1 shl 16)
        while (true) { val n = input.read(buf); if (n < 0) break; md.update(buf, 0, n) }
        return md.digest().joinToString("") { "%02x".format(it) }
    }

    private fun x25519(priv: ByteArray, pub: ByteArray): ByteArray {
        val a = X25519Agreement(); a.init(X25519PrivateKeyParameters(priv, 0))
        val out = ByteArray(32)
        a.calculateAgreement(X25519PublicKeyParameters(checkPub(pub), 0), out, 0)
        return out
    }

    private fun checkPub(pub: ByteArray): ByteArray {
        if (pub.size != 32) throw CryptoException("bad public key")
        return pub
    }

    private fun hkdf(shared: ByteArray, salt: ByteArray, label: String): ByteArray {
        val g = HKDFBytesGenerator(SHA256Digest())
        g.init(HKDFParameters(shared, salt, INFO + label.toByteArray()))
        val out = ByteArray(32); g.generateBytes(out, 0, 32); return out
    }

    private fun gcm(mode: Int, key: ByteArray, nonce: ByteArray, aad: ByteArray, data: ByteArray): ByteArray {
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        c.init(mode, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        c.updateAAD(aad)
        return c.doFinal(data)
    }

    private fun fileNonce(i: Long): ByteArray = ByteBuffer.allocate(12).putLong(i).putInt(0).array()
    private fun fileAad(i: Long, last: Boolean) = "file:$i:${if (last) 1 else 0}".toByteArray()

    class Sealer(senderPriv: ByteArray, recipientPub: ByteArray) {
        val header: JSONObject
        private val kMeta: ByteArray
        private val kFile: ByteArray

        init {
            val eph = X25519PrivateKeyParameters(rng)
            val salt = ByteArray(16).also { rng.nextBytes(it) }
            val shared = x25519(eph.encoded, recipientPub) + x25519(senderPriv, recipientPub)
            kMeta = hkdf(shared, salt, "meta")
            kFile = hkdf(shared, salt, "file")
            header = JSONObject().put("v", VERSION).put("salt", b64e(salt)).put("eph", b64e(eph.generatePublicKey().encoded))
        }

        fun sealMeta(obj: JSONObject): String {
            val nonce = ByteArray(12).also { rng.nextBytes(it) }
            val ct = gcm(Cipher.ENCRYPT_MODE, kMeta, nonce, "meta".toByteArray(), obj.toString().toByteArray())
            return b64e(nonce + ct)
        }

        fun sealStream(input: InputStream, size: Long, out: OutputStream) {
            var done = 0L; var i = 0L
            val buf = ByteArray(CHUNK)
            while (true) {
                val n = readFull(input, buf)
                done += n
                val last = done >= size
                val ct = gcm(Cipher.ENCRYPT_MODE, kFile, fileNonce(i), fileAad(i, last), buf.copyOf(n))
                out.write(ByteBuffer.allocate(4).putInt(ct.size).array()); out.write(ct)
                i++
                if (last) break
            }
        }
    }

    class Opener(recipientPriv: ByteArray, senderPub: ByteArray, header: JSONObject) {
        private val kMeta: ByteArray
        private val kFile: ByteArray

        init {
            if (header.optInt("v", 0) != VERSION) throw CryptoException("unsupported envelope version")
            val salt = b64d(header.optString("salt"))
            val eph = b64d(header.optString("eph"))
            val shared = x25519(recipientPriv, eph) + x25519(recipientPriv, senderPub)
            kMeta = hkdf(shared, salt, "meta")
            kFile = hkdf(shared, salt, "file")
        }

        fun openMeta(blob: String): JSONObject {
            val raw = b64d(blob)
            if (raw.size < 28) throw CryptoException("metadata blob too short")
            val pt = try {
                gcm(Cipher.DECRYPT_MODE, kMeta, raw.copyOfRange(0, 12), "meta".toByteArray(), raw.copyOfRange(12, raw.size))
            } catch (e: Exception) { throw CryptoException("metadata failed to authenticate") }
            return try { JSONObject(String(pt)) } catch (e: Exception) { throw CryptoException("metadata is not JSON") }
        }

        fun openStream(input: InputStream, out: OutputStream) {
            var i = 0L; var last = false
            val hdr = ByteArray(4)
            while (!last) {
                if (readFull(input, hdr) < 4) throw CryptoException("file stream truncated")
                val n = ByteBuffer.wrap(hdr).int
                if (n < 16 || n > CHUNK + 16) throw CryptoException("file frame has an impossible size")
                val ct = ByteArray(n)
                if (readFull(input, ct) < n) throw CryptoException("file stream truncated")
                var pt: ByteArray? = null
                for (flag in listOf(false, true)) {
                    try { pt = gcm(Cipher.DECRYPT_MODE, kFile, fileNonce(i), fileAad(i, flag), ct); last = flag; break }
                    catch (e: Exception) { /* try the other flag */ }
                }
                out.write(pt ?: throw CryptoException("file chunk $i failed to authenticate"))
                i++
            }
            if (input.read() != -1) throw CryptoException("data after the final chunk")
        }

        fun openBytes(data: ByteArray): ByteArray {
            val out = java.io.ByteArrayOutputStream(); openStream(data.inputStream(), out); return out.toByteArray()
        }
    }

    /** Fill buf as far as the stream allows; returns bytes read (0 at EOF). */
    private fun readFull(input: InputStream, buf: ByteArray): Int {
        var got = 0
        while (got < buf.size) {
            val n = input.read(buf, got, buf.size - got)
            if (n < 0) break
            got += n
        }
        return got
    }

    fun checkFreshness(meta: JSONObject, myId: String) {
        val ts = meta.optDouble("ts", 0.0)
        if (kotlin.math.abs(System.currentTimeMillis() / 1000.0 - ts) > MAX_SKEW) throw CryptoException("envelope is stale")
        val to = meta.optString("to", "")
        if (to.isNotEmpty() && to != myId) throw CryptoException("envelope is addressed to another device")
    }

    // ── pairing codes ────────────────────────────────────────────────────
    data class Pairing(val name: String, val url: String, val pub: ByteArray, val key: String, val id: String)

    fun makePairingCode(name: String, url: String, pub: ByteArray, keyIn: String, id: String): String {
        val body = JSONObject().put("v", VERSION).put("name", name).put("url", url).put("pub", b64e(pub)).put("key", keyIn).put("id", id)
        return "fs1." + b64e(body.toString().toByteArray())
    }

    fun parsePairingCode(code: String): Pairing {
        val c = code.trim()
        if (!c.startsWith("fs1.")) throw IllegalArgumentException("not a family_share pairing code")
        val d = try { JSONObject(String(b64d(c.substring(4)))) } catch (e: Exception) { throw IllegalArgumentException("pairing code is corrupt") }
        val name = d.optString("name"); val pub = d.optString("pub"); val key = d.optString("key")
        if (name.isEmpty() || pub.isEmpty() || key.isEmpty()) throw IllegalArgumentException("pairing code is missing fields")
        return Pairing(name, d.optString("url"), checkPub(b64d(pub)), key, d.optString("id"))
    }
}

File: modules/family_share/crypto.py
"""
family_share: end-to-end encryption.
======================================================================
Every photo and every byte of metadata is encrypted on the SENDING instance
and decrypted only on the RECEIVING instance. Nothing in between — the
apartment router, the ISP, whoever is on the wifi, a reverse proxy — sees
more than ciphertext plus a peer name and an envelope header. TLS is still
worth having (it hides the peer name and sizes) but nothing here relies on it.

Scheme (one envelope per push / revoke):

  * each instance owns a long-lived X25519 key pair; peers exchange public
    keys once, inside the pairing code, and pin them;
  * per message the sender makes an EPHEMERAL X25519 key and derives
        shared = X25519(eph, recipient_pub) || X25519(sender_static, recipient_pub)
    so a message is readable only by the recipient's private key (and gains
    forward secrecy from the ephemeral half) AND could only have been built
    by the holder of the sender's private key (the static half). A stolen
    peer key or a replayed header cannot forge a push;
  * HKDF-SHA256 turns `shared` into separate AES-256 keys for the metadata
    blob and the file stream (never the same key for two nonce spaces);
  * AES-256-GCM authenticates everything. The file is a STREAM of 1 MiB
    chunks, each with its index and a last-chunk flag bound into the AAD, so
    chunks can't be dropped, reordered, truncated or spliced;
  * the encrypted metadata carries a timestamp and the recipient's instance
    id; the receiver rejects stale envelopes and ones meant for someone else.

The private key lives in app_config.json next to every other secret the app
already keeps (LDAP bind password, session keys). Rotating it means re-pairing.
"""

import base64
import io
import json
import os
import struct
import time

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey, X25519PublicKey
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

VERSION = 1
CHUNK = 1 << 20                 # 1 MiB plaintext per frame
MAX_SKEW = 15 * 60              # seconds an envelope stays acceptable
_INFO = b"family_share/v1/"


class CryptoError(Exception):
    pass


# ── keys ────────────────────────────────────────────────────────────────────
def b64e(b):
    return base64.urlsafe_b64encode(b).decode().rstrip("=")


def b64d(s):
    s = str(s or "")
    return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))


def generate_private_key():
    """-> base64 raw 32-byte private key."""
    return b64e(X25519PrivateKey.generate().private_bytes(
        serialization.Encoding.Raw, serialization.PrivateFormat.Raw, serialization.NoEncryption()))


def public_key(priv_b64):
    return b64e(_priv(priv_b64).public_key().public_bytes(
        serialization.Encoding.Raw, serialization.PublicFormat.Raw))


def fingerprint(pub_b64):
    """Short human-checkable form of a public key (first 16 hex of SHA-256)."""
    h = hashes.Hash(hashes.SHA256()); h.update(b64d(pub_b64))
    hexd = h.finalize().hex()[:16]
    return " ".join(hexd[i:i + 4] for i in range(0, 16, 4))


def _priv(b64):
    try:
        return X25519PrivateKey.from_private_bytes(b64d(b64))
    except Exception as e:
        raise CryptoError(f"bad private key: {e}")


def _pub(b64):
    try:
        raw = b64d(b64)
        if len(raw) != 32:
            raise ValueError("wrong length")
        return X25519PublicKey.from_public_bytes(raw)
    except Exception as e:
        raise CryptoError(f"bad public key: {e}")


def _derive(shared, salt, label):
    return HKDF(algorithm=hashes.SHA256(), length=32, salt=salt,
                info=_INFO + label).derive(shared)


# ── sender ──────────────────────────────────────────────────────────────────
class Sealer:
    """One envelope: seal_meta() once, then stream a file through seal_file()."""

    def __init__(self, sender_priv_b64, recipient_pub_b64):
        eph = X25519PrivateKey.generate()
        rpub = _pub(recipient_pub_b64)
        spriv = _priv(sender_priv_b64)
        self.salt = os.urandom(16)
        shared = eph.exchange(rpub) + spriv.exchange(rpub)
        self._k_meta = _derive(shared, self.salt, b"meta")
        self._k_file = _derive(shared, self.salt, b"file")
        self.header = {"v": VERSION, "salt": b64e(self.salt),
                       "eph": b64e(eph.public_key().public_bytes(
                           serialization.Encoding.Raw, serialization.PublicFormat.Raw))}

    def seal_meta(self, obj):
        nonce = os.urandom(12)
        ct = AESGCM(self._k_meta).encrypt(nonce, json.dumps(obj).encode("utf-8"), b"meta")
        return b64e(nonce + ct)

    def seal_file(self, src_path, dst_path):
        """Write the framed ciphertext stream of src_path to dst_path."""
        with open(src_path, "rb") as fin, open(dst_path, "wb") as fout:
            self.seal_stream(fin, os.path.getsize(src_path), fout)
        return dst_path

    def seal_stream(self, fin, size, fout):
        """Frame `size` bytes read from fin into fout (see module docstring)."""
        aes = AESGCM(self._k_file)
        done = 0; i = 0
        while True:
            data = fin.read(CHUNK)
            done += len(data)
            last = done >= size
            nonce = struct.pack(">Q", i) + b"\0\0\0\0"
            ct = aes.encrypt(nonce, data, _file_aad(i, last))
            fout.write(struct.pack(">I", len(ct)) + ct)
            i += 1
            if last:
                break

    def iter_frames(self, path, size):
        """Generator of sealed frames for a streamed response, made as they are
        sent so a video never sits in RAM."""
        aes = AESGCM(self._k_file)
        done = 0; i = 0
        with open(path, "rb") as fin:
            while True:
                data = fin.read(CHUNK)
                done += len(data)
                last = done >= size
                nonce = struct.pack(">Q", i) + b"\0\0\0\0"
                ct = aes.encrypt(nonce, data, _file_aad(i, last))
                yield struct.pack(">I", len(ct)) + ct
                i += 1
                if last:
                    return

    def seal_bytes(self, data):
        """Framed ciphertext of an in-memory blob (a thumbnail, a listing)."""
        out = io.BytesIO()
        self.seal_stream(io.BytesIO(data), len(data), out)
        return out.getvalue()


def _file_aad(i, last):
    return b"file:%d:%d" % (i, 1 if last else 0)


# ── receiver ────────────────────────────────────────────────────────────────
class Opener:
    def __init__(self, recipient_priv_b64, sender_pub_b64, header):
        try:
            if int(header.get("v", 0)) != VERSION:
                raise CryptoError("unsupported envelope version")
            self.salt = b64d(header["salt"])
            eph = _pub(header["eph"])
        except CryptoError:
            raise
        except Exception as e:
            raise CryptoError(f"bad envelope header: {e}")
        rpriv = _priv(recipient_priv_b64)
        shared = rpriv.exchange(eph) + rpriv.exchange(_pub(sender_pub_b64))
        self._k_meta = _derive(shared, self.salt, b"meta")
        self._k_file = _derive(shared, self.salt, b"file")

    def open_meta(self, blob_b64):
        raw = b64d(blob_b64)
        if len(raw) < 12 + 16:
            raise CryptoError("metadata blob too short")
        try:
            pt = AESGCM(self._k_meta).decrypt(raw[:12], raw[12:], b"meta")
        except Exception:
            raise CryptoError("metadata failed to authenticate: wrong sender key or tampered")
        try:
            obj = json.loads(pt.decode("utf-8"))
        except Exception:
            raise CryptoError("metadata is not JSON")
        if not isinstance(obj, dict):
            raise CryptoError("metadata is not an object")
        return obj

    def open_file(self, stream, dst_path):
        """Decrypt a framed stream (a file-like with .read) into dst_path."""
        with open(dst_path, "wb") as fout:
            self.open_stream(stream, fout)
        return dst_path

    def open_bytes(self, data):
        out = io.BytesIO()
        self.open_stream(io.BytesIO(data), out)
        return out.getvalue()

    def open_stream(self, stream, fout):
        aes = AESGCM(self._k_file)
        i = 0; last = False
        if True:
            while not last:
                hdr = _read_exact(stream, 4)
                if hdr is None:
                    raise CryptoError("file stream truncated")
                (n,) = struct.unpack(">I", hdr)
                if n < 16 or n > CHUNK + 16:
                    raise CryptoError("file frame has an impossible size")
                ct = _read_exact(stream, n)
                if ct is None:
                    raise CryptoError("file stream truncated")
                nonce = struct.pack(">Q", i) + b"\0\0\0\0"
                pt = None
                for flag in (False, True):
                    try:
                        pt = aes.decrypt(nonce, ct, _file_aad(i, flag)); last = flag; break
                    except Exception:
                        continue
                if pt is None:
                    raise CryptoError(f"file chunk {i} failed to authenticate")
                fout.write(pt)
                i += 1
            if stream.read(1):
                raise CryptoError("data after the final chunk")


def _read_exact(stream, n):
    buf = b""
    while len(buf) < n:
        part = stream.read(n - len(buf))
        if not part:
            return None
        buf += part
    return buf


def check_freshness(meta, my_id):
    """The decrypted metadata must be recent and addressed to us."""
    ts = float(meta.get("ts") or 0)
    if abs(time.time() - ts) > MAX_SKEW:
        raise CryptoError("envelope is stale (clock skew over 15 minutes, or a replay)")
    if meta.get("to") and meta["to"] != my_id:
        raise CryptoError("envelope is addressed to another instance")


# ── pairing code ────────────────────────────────────────────────────────────
def make_pairing_code(name, url, pub_b64, key_in, instance_id=""):
    """One string to hand the other side: who I am, where I am, my public key,
    the secret they must send me."""
    body = json.dumps({"v": VERSION, "name": name, "url": url, "pub": pub_b64, "key": key_in,
                       "id": instance_id}, separators=(",", ":")).encode()
    return "fs1." + b64e(body)


def parse_pairing_code(code):
    code = str(code or "").strip()
    if not code.startswith("fs1."):
        raise ValueError("not a family_share pairing code")
    try:
        d = json.loads(b64d(code[4:]).decode())
    except Exception:
        raise ValueError("pairing code is corrupt")
    if not isinstance(d, dict) or not d.get("name") or not d.get("pub") or not d.get("key"):
        raise ValueError("pairing code is missing fields")
    _pub(d["pub"])       # validates
    return {"name": str(d["name"])[:64], "url": str(d.get("url") or "")[:512],
            "pub_key": str(d["pub"]), "key_out": str(d["key"]),
            "instance_id": str(d.get("id") or "")[:64]}
